File: sdks/typescript/src/http.ts
/** Fetch-based transport. No dependencies: Node 20+ has global fetch, Node 22+
 * has global WebSocket (used by the exec stream). */

import { IgnitionError, errorFromResponse } from "./errors.ts";

const USER_AGENT = "ignition-sandbox-ts/0.1";

export interface TransportOptions {
  server?: string;
  token?: string;
  project?: string;
  timeoutMs?: number;
}

function env(name: string): string {
  return (globalThis as any).process?.env?.[name] ?? "";
}

export class Transport {
  readonly server: string;
  readonly token: string;
  readonly project: string;
  readonly timeoutMs: number;

  constructor(opts: TransportOptions) {
    this.server = (opts.server || env("IGNITION_SERVER")).replace(/\/+$/, "");
    this.token = opts.token || env("IGNITION_TOKEN");
    this.project = opts.project || env("IGNITION_PROJECT");
    this.timeoutMs = opts.timeoutMs ?? 30_000;
    if (!this.server) {
      throw new IgnitionError("no server configured: pass server or set IGNITION_SERVER");
    }
  }

  projectPath(suffix: string): string {
    if (!this.project) {
      throw new IgnitionError("no project configured: pass project or set IGNITION_PROJECT");
    }
    return `/v1/projects/${this.project}${suffix}`;
  }

  private headers(extra: Record<string, string> = {}): Record<string, string> {
    const h: Record<string, string> = { "User-Agent": USER_AGENT, ...extra };
    if (this.token) h.Authorization = `Bearer ${this.token}`;
    return h;
  }

  async request<T = unknown>(
    method: string,
    path: string,
    opts: {
      body?: unknown;
      idempotent?: boolean;
      idempotencyKey?: string;
      query?: Record<string, string | undefined>;
    } = {},
  ): Promise<T> {
    let url = this.server + path;
    if (opts.query) {
      const qs = Object.entries(opts.query)
        .filter(([, v]) => v !== undefined && v !== "")
        .map(([k, v]) => `${k}=${encodeURIComponent(v as string)}`)
        .join("&");
      if (qs) url += `?${qs}`;
    }

    const headers = this.headers({ Accept: "application/json" });
    let body: string | undefined;
    if (opts.body !== undefined) {
      body = JSON.stringify(opts.body);
      headers["Content-Type"] = "application/json";
    }
    if (opts.idempotent) {
      headers["Idempotency-Key"] = opts.idempotencyKey || `tsdk-${crypto.randomUUID()}`;
    }

    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), this.timeoutMs);
    let resp: Response;
    try {
      resp = await fetch(url, { method, headers, body, signal: ac.signal });
    } catch (e) {
      throw new IgnitionError(`${method} ${path}: ${(e as Error).message}`);
    } finally {
      clearTimeout(timer);
    }

    const text = await resp.text();
    if (!resp.ok) {
      throw errorFromResponse(resp.status, text, resp.headers.get("x-request-id") ?? "");
    }
    return (text ? JSON.parse(text) : undefined) as T;
  }

  get<T = unknown>(path: string, opts?: Parameters<Transport["request"]>[2]): Promise<T> {
    return this.request<T>("GET", path, opts);
  }

  post<T = unknown>(path: string, body?: unknown, opts: Parameters<Transport["request"]>[2] = {}): Promise<T> {
    return this.request<T>("POST", path, { ...opts, body });
  }

  /** Yield each JSON snapshot from an SSE `:watch` stream. */
  async *sse(path: string, lastEventId = ""): AsyncGenerator<any> {
    const headers = this.headers({ Accept: "text/event-stream" });
    if (lastEventId) headers["Last-Event-ID"] = lastEventId;
    const resp = await fetch(this.server + path, { method: "GET", headers });
    if (!resp.ok) {
      throw errorFromResponse(resp.status, await resp.text(), resp.headers.get("x-request-id") ?? "");
    }
    const reader = resp.body!.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) return;
        buf += decoder.decode(value, { stream: true });
        let nl: number;
        while ((nl = buf.indexOf("\n")) >= 0) {
          const line = buf.slice(0, nl).replace(/\r$/, "");
          buf = buf.slice(nl + 1);
          if (line.startsWith("data: ")) {
            try {
              yield JSON.parse(line.slice(6));
            } catch {
              /* ignore keepalive / partial */
            }
          }
        }
      }
    } finally {
      try {
        await reader.cancel();
      } catch {
        /* already closed */
      }
    }
  }
}

File: sdks/python/ignition_sandbox/_http.py
"""Minimal HTTP transport over the standard library.

The SDK deliberately has no runtime dependencies: this wraps ``urllib.request``
with the small amount of behavior the Ignition API needs — bearer auth, an
``Idempotency-Key`` on every mutation, stable error mapping, and a line reader
for the Server-Sent-Events ``:watch`` streams.
"""

from __future__ import annotations

import json
import os
import uuid
from typing import Any, Iterator
from urllib import error as urlerror
from urllib import request as urlrequest

from .errors import IgnitionError, from_response

_USER_AGENT = "ignition-sandbox-python/0.1"


class Transport:
    def __init__(self, server: str, token: str, project: str, timeout: float) -> None:
        if not server:
            raise IgnitionError(
                "no server configured: pass server=... or set IGNITION_SERVER"
            )
        self.server = server.rstrip("/")
        self.token = token
        self.project = project
        self.timeout = timeout

    # -- path helpers ------------------------------------------------------
    def _url(self, path: str) -> str:
        return self.server + path

    def project_path(self, suffix: str) -> str:
        if not self.project:
            raise IgnitionError(
                "no project configured: pass project=... or set IGNITION_PROJECT"
            )
        return f"/v1/projects/{self.project}{suffix}"

    # -- requests --------------------------------------------------------
    def request(
        self,
        method: str,
        path: str,
        body: Any | None = None,
        *,
        idempotent: bool = False,
        idempotency_key: str | None = None,
        query: dict[str, str] | None = None,
    ) -> Any:
        url = self._url(path)
        if query:
            pairs = "&".join(f"{k}={v}" for k, v in query.items() if v not in (None, ""))
            if pairs:
                url += "?" + pairs

        data = None
        headers = {"Accept": "application/json", "User-Agent": _USER_AGENT}
        if body is not None:
            data = json.dumps(body).encode()
            headers["Content-Type"] = "application/json"
        if self.token:
            headers["Authorization"] = "Bearer " + self.token
        if idempotent:
            headers["Idempotency-Key"] = idempotency_key or ("isdk-" + uuid.uuid4().hex)

        req = urlrequest.Request(url, data=data, method=method, headers=headers)
        try:
            with urlrequest.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
                if not raw:
                    return None
                return json.loads(raw)
        except urlerror.HTTPError as exc:  # non-2xx
            raw = exc.read()
            rid = exc.headers.get("X-Request-Id", "") if exc.headers else ""
            raise from_response(exc.code, raw, rid) from None
        except urlerror.URLError as exc:
            raise IgnitionError(f"{method} {path}: {exc.reason}") from None

    def get(self, path: str, **kw: Any) -> Any:
        return self.request("GET", path, **kw)

    def post(self, path: str, body: Any | None = None, **kw: Any) -> Any:
        return self.request("POST", path, body, **kw)

    # -- SSE ------------------------------------------------------------
    def sse(self, path: str, last_event_id: str = "") -> Iterator[dict]:
        """Yield each JSON snapshot from an SSE ``:watch`` stream.

        The connection is closed when the caller stops iterating.
        """
        url = self._url(path)
        headers = {"Accept": "text/event-stream", "User-Agent": _USER_AGENT}
        if self.token:
            headers["Authorization"] = "Bearer " + self.token
        if last_event_id:
            headers["Last-Event-ID"] = last_event_id
        req = urlrequest.Request(url, method="GET", headers=headers)
        try:
            resp = urlrequest.urlopen(req, timeout=None)
        except urlerror.HTTPError as exc:
            raw = exc.read()
            rid = exc.headers.get("X-Request-Id", "") if exc.headers else ""
            raise from_response(exc.code, raw, rid) from None
        with resp:
            for line in resp:
                text = line.decode("utf-8", "replace").rstrip("\n")
                if text.startswith("data: "):
                    payload = text[6:]
                    try:
                        yield json.loads(payload)
                    except json.JSONDecodeError:
                        continue


def env_default(explicit: str | None, *names: str) -> str:
    if explicit:
        return explicit
    for n in names:
        v = os.environ.get(n)
        if v:
            return v
    return ""
